File: ros2_ws/rti_workspace/5.2.3/examples/distributed_logger/java/hello_log4j_logging/src/com/rti/hello/HelloPublisher.java
// ****************************************************************************
//         (c) Copyright, Real-Time Innovations, All rights reserved.       
//                                                                          
//         Permission to modify and use for internal purposes granted.      
// This software is provided "as is", without warranty, express or implied. 
//                                                                          
// ****************************************************************************

package com.rti.hello;


import org.apache.log4j.Logger;

import com.rti.dds.domain.DomainParticipant;
import com.rti.dds.infrastructure.InstanceHandle_t;
import com.rti.dds.infrastructure.RETCODE_ERROR;
import com.rti.dds.infrastructure.StatusKind;
import com.rti.dds.publication.Publisher;
import com.rti.dds.topic.Topic;
import com.rti.hello.idl.HelloWorld;
import com.rti.hello.idl.HelloWorldDataWriter;



//****************************************************************************
/**
 * Manage the data publication for the Hello example 
 */
public class HelloPublisher {
    // get a logger from the java.util.logging package
    private static final Logger logger = Logger.getLogger(
            HelloPublisher.class);

    // Maximum consecutive write error after stopping the write process
    private static final int MAX_CONSECUTIVE_WRITE_ERROR = 5;
    
    private HelloWorldDataWriter _theDataWriter;
    private int                  _theDataSize;
    private int                  _theVerbose;
    private int                  _theSampleCount = 0;

    // ************************************************************************
    /**
     * Builds a new HelloPublisher object and initialize the inner DDS 
     * entities.
     * This method only creates the entities, but it does not actually write
     * the data.
     * To write the data, use the method run().
     *  
     * @param participant the DDS Domain participant
     * @param topic       the DDS topic 
     * @param verbose     an integer to control the output verbosity
     * @param dataSize    the size of the payload data
     */
    public HelloPublisher(DomainParticipant participant,
            Topic topic,
            int verbose,
            int dataSize,
            int sampleCount) {
        _theDataSize = dataSize;
        _theVerbose = verbose;
        _theSampleCount = sampleCount;
        
        // Create the publisher. 
        // Just like the participant, if you want to customize the publisher QoS,
        // use DomainParticipant.get_default_publisher_qos() to 
        // initialize a local copy of the default QoS, modify them, then
        // use them in the creation call below instead of 
        //     DomainParticipant.PUBLISHER_QOS_DEFAULT.
        // For more info on publisher API, see:
        //     $NDDSHOME/doc/html/api_java/group__DDSPublisherModule.html
        //
        if (verbose > 0) {
            logger.info("Creating publisher...");
        }
        Publisher publisher = participant.create_publisher(
                    DomainParticipant.PUBLISHER_QOS_DEFAULT, 
                    null,           // listener
                    StatusKind.STATUS_MASK_NONE);
        if (publisher == null) {
            logger.error("! Unable to create DDS Publisher");
            throw new RuntimeException("HelloPublisher creation failed");
        }            
        
        // Creates the DDS Data writer. 
        // Just like before, if you want to customize the writer QoS,
        // use Publisher.get_default_datawriter_qos() to 
        // initialize a local copy of the default QoS, modify them, then
        // use them in the creation call below instead of 
        // DATAWRITER_QOS_DEFAULT.
        // For more data writer API info, see:
        //     $NDDSHOME/doc/html/api_c/group__DDSWriterModule.html
        //
        if (verbose > 0) {
            logger.info("Creating the data writer...");
        }
        _theDataWriter = (HelloWorldDataWriter)
            publisher.create_datawriter(topic, 
                        Publisher.DATAWRITER_QOS_DEFAULT,
                        null,           // listener
                        StatusKind.STATUS_MASK_NONE);
        if (_theDataWriter == null) {
            logger.error("! Unable to create DDS data writer\n");
            return;
        }
    }
    
    // ************************************************************************
    /**
     * Writes samples to DDS. This method never returns.
     */
    void writeSamples() {
        HelloWorld instance = new HelloWorld();
        
        // Initialize the data
        instance.prefix = "Hello world";
        instance.payload.setSize(_theDataSize);
        byte arr[] = (byte [])instance.payload.getPrimitiveArray();
        for (int i = 0; i < _theDataSize; ++i) {
            arr[i] = (byte)(i % 0xff);
        }
        
        logger.info("Sending data...");
        int consecutiveErrors = 0;
        for (instance.sample_id = 0;; ++(instance.sample_id) ) {
            if (instance.sample_id % 1000 == 0) {
                logger.info("Sending sample id : " + instance.sample_id);
            }
            try {
                _theDataWriter.write(instance, InstanceHandle_t.HANDLE_NIL);
            } catch(RETCODE_ERROR e) {
                logger.error("! Write error " + 
                        e.getClass().toString() + ": " + e.getMessage());
                if (++consecutiveErrors > MAX_CONSECUTIVE_WRITE_ERROR) {
                    logger.error("! Reached maximum number of failure, " +
                        "stopping writer...");
                    return;
                }
            }
            // Always clear the error count in case of successful write
            consecutiveErrors = 0;
            if (_theVerbose > 0 && (instance.sample_id % 10000) == 0) {
                logger.info("Sent " + 
                        Integer.toString(instance.sample_id) + 
                        " samples...");
            }
            if (_theSampleCount != 0 && (instance.sample_id >= _theSampleCount)) {
                logger.info("\nSent " + 
                        Integer.toString(instance.sample_id) + 
                        " samples.");
                return;
            }
        }
    }
}
 

File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/c/hello_idl/src/HelloPublisher.c
/*****************************************************************************/
/*         (c) Copyright, Real-Time Innovations, All rights reserved.        */
/*                                                                           */
/*         Permission to modify and use for internal purposes granted.       */
/* This software is provided "as is", without warranty, express or implied.  */
/*                                                                           */
/*****************************************************************************/


#include "Hello.h"
#include "HelloPublisher.h"
#include "HelloWorld.h"
#include "HelloWorldSupport.h"


/* Defines the maximum number of consecutive write errors. */
#define MAX_CONSECUTIVE_WRITE_ERROR         10



/*****************************************************************************/
/* start_publisher                                                           */
/*                                                                           */
/* Creates the DDS publisher and data writers, then start sending the data.  */
/*                                                                           */
/* Input:                                                                    */
/*   participant: the DDS domain participant to use                          */
/*   topic: the DDS topic to use for publishing                              */
/*   dataSise: size of the payload buffer to send                            */
/*                                                                           */
/* Returns:                                                                  */
/*   The boolean value DDS_BOOLEAN_TRUE if an error occurred.                */
/*                                                                           */
/*****************************************************************************/
DDS_Boolean start_publisher(
                        DDS_DomainParticipant *participant,
                        DDS_Topic *topic,
                        DDS_Boolean verbose,
                        DDS_Long data_size,
                        DDS_Long sample_count) {

    DDS_Publisher *publisher             = NULL;
    DDS_DataWriter *data_writer          = NULL;
    HelloWorldDataWriter *hello_writer   = NULL;
    HelloWorld *instance                 = NULL;
    DDS_Octet * payload_buffer           = NULL;
    DDS_Boolean return_value             = DDS_BOOLEAN_FALSE;
    DDS_Long i;
    DDS_ReturnCode_t rc;
    struct DDS_Duration_t send_period = {0,4}; /* time (sec, usec) to pause between bursts of 10,000 samples */
    struct DDS_Duration_t disc_period = {1,0};
    
    /* Create the publisher. 
     * Just like the participant, if you want to customize the publisher QoS,
     * use DDS_DomainParticipant_get_default_publisher_qos() to 
     * initialize a local copy of the default QoS, modify them, then
     * use them in the creation call below instead of 
     * DDS_PUBLISHER_QOS_DEFAULT.
     * For more info on publisher API, see:
     *     $NDDSHOME/doc/html/api_c/group__DDSPublisherModule.html
     */
    if (verbose) {
        printf("Creating the publisher...\n");
    }
    publisher = DDS_DomainParticipant_create_publisher(
                        participant,
                        &DDS_PUBLISHER_QOS_DEFAULT,
                        NULL,           /* listener */
                        DDS_STATUS_MASK_NONE);
    if (publisher == NULL) {
        fprintf(stderr, "! Unable to create DDS Publisher\n");
        goto exitFn;
    }

    /* Creates the DDS Data writer. 
     * Just like before, if you want to customize the writer QoS,
     * use DDS_Publisher_get_default_datawriter_qos() to 
     * initialize a local copy of the default QoS, modify them, then
     * use them in the creation call below instead of 
     * DDS_DATAWRITER_QOS_DEFAULT.
     * For more data writer API info, see:
     *     $NDDSHOME/doc/html/api_c/group__DDSWriterModule.html
     */
    if (verbose) {
        printf("Creating the data writer...\n");
    }
    data_writer = DDS_Publisher_create_datawriter(
                        publisher, 
                        topic,
                        &DDS_DATAWRITER_QOS_DEFAULT,
                        NULL,           /* listener */
                        DDS_STATUS_MASK_NONE);
    if (data_writer == NULL) {
        fprintf(stderr, "! Unable to create DDS data writer\n");
        goto exitFn;
    }

    /* wait for discovery */
    NDDS_Utility_sleep(&disc_period);

    /* The following narrow function should never fail, as it performs 
     * only a safe cast of the generic data writer into a specific
     * HelloWorldDataWriter.
     * In our case the following cast would do the same job:
     *     hello_writer = (HelloWorldDataWriter *)data_writer;
     */
    hello_writer = HelloWorldDataWriter_narrow(data_writer);
    if (hello_writer == NULL) {
        fprintf(stderr, "! Unable to narrow data writer into HelloWorldDataWriter\n");
        goto exitFn;
    }

    /* Everything is ready, start writing the data.
     * The create data function below simply allocates on the heap an 
     * instance of the data to send.
     * Alternatively you can allocate the HelloWorld instance on the
     * struct, and call the HelloWorld_initialize_ex() function
     * to initialize the field to a known value.
     * For more information on the user data type support, see:
     *     $NDDSHOME/doc/html/api_c/group__DDSUserDataModule.html
     *
     * In this case we create an instance of the data on the heap, but
     * we tell the HelloWorld_create_data_ex to skip allocation of the
     * inner buffers (in our case, the 'prefix' string).
     */
    instance = HelloWorldTypeSupport_create_data_ex(
                        DDS_BOOLEAN_FALSE);
    if (instance == NULL) {
        fprintf(stderr, "! Unable to create an instance of the data\n");
        fprintf(stderr, "! This problem most likely is caused by "
                        "out of memory\n");
        goto exitFn;
    }

    /* Sets the length of the sequence we are sending */
    if (!DDS_OctetSeq_ensure_length(&instance->payload, 
                        data_size, 
                        HELLODDS_MAX_PAYLOAD_SIZE)) {
        fprintf(stderr, "! Unable to set payload size to %d\n", data_size);
        fprintf(stderr, "! Perhaps you are using a value too big "
                        "(max allowed size is %d)\n", 
                        HELLODDS_MAX_PAYLOAD_SIZE);
        goto exitFn;
    }
    
    /* The payload_buffer has been already allocated during the previous call
     * now we get a pointer to the buffer used to manage the sequence
     * so we can address directly the bytes.
     * For more information on sequence, see:
     *     $NDDSHOME/doc/html/api_c/group__DDSSequenceModule.html
     *
     */
    payload_buffer = DDS_OctetSeq_get_contiguous_buffer(&instance->payload);
    

    /* Fill up the buffer with some valid data:
     * The content of the payload buffer is a progressive series of incremental
     * numbers, starting from 0.
     */
    for (i = 0; i < data_size; ++i) {
        payload_buffer[i] = (i & 0xff);
    }
    
    sprintf(instance->prefix, "Hello world");
    
    /* Send the data! */
    printf("Sending data...\n");
    i = 0; /* Use 'i' to keep track of consecutive write errors */
    for (instance->sampleId = 0;; ++(instance->sampleId) ) {
        rc = HelloWorldDataWriter_write(
                        hello_writer, 
                        instance, 
                        &DDS_HANDLE_NIL);
        if (rc != DDS_RETCODE_OK) {
            fprintf(stderr, "! Write error %d\n", rc);
            if (++i > MAX_CONSECUTIVE_WRITE_ERROR) {
                fprintf(stderr, "! Reached maximum number of failure, "
                       "stopping writer...\n");
                goto exitFn;
            }
        } else {
            i = 0;  /* Always clear the error count after a successful write */
        }
        if ((instance->sampleId % 10000) == 0) {
            if (verbose) {
                printf("Sent %d samples...\n", instance->sampleId);
            }
        NDDS_Utility_sleep(&send_period);
        }
        if (sample_count != 0 && instance->sampleId >= sample_count) {
            printf("\nSent %d samples.\n", instance->sampleId);
            break;
        }
    }
    
    /* This code is reached when sampleCount is set */
    return_value = DDS_BOOLEAN_TRUE;

exitFn:
    if (instance != NULL) {
        HelloWorldTypeSupport_delete_data_ex(instance,
                        DDS_BOOLEAN_FALSE);
        instance = NULL;
    }
    
    return return_value;
}

